System Status performance improvements

joshbetz opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

1. Ensures we don't calculate unnecessary status fields
2. Caches theme info

### How to test the changes in this Pull Request:

1. Make requests to /wp-json/wc/v3/system_status?_fields=theme and /wp-json/wc/v3/system_status?_fields=environment
2. Observe that the response times are approximately the same
3. Checkout this PR
4. Make 2 requests to /wp-json/wc/v3/system_status?_fields=theme
5. The second request should be quite a bit faster (it was ~600ms in my testing)
6. Switch themes
7. Observe that /wp-json/wc/v3/system_status?_fields=theme shows the correct info

If you want to confirm that we are indeed only calculating the necessary data, you could add some logging to get_theme_info and send a request to /wp-json/wc/v3/system_status?_fields=environment to observe that we no longer call get_theme_info in that case.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [x] Have you created a changelog file by running `pnpm nx affected --target=changelog`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshbetz committed:
API: System Status: Only load necessary data

The system status endpoint supports the global `_fields` query parameter
for limiting which fields are returned. That is, if you send a request
with a query string of `?_fields=environment`, only the environment
property will be returned. However, we still calculate every property.

This change ensure that we only gather the data that is necessary for
the response. Since some of this data can be somewhat slow (like reading
from the disk in the case of get_theme_info), this could have a pretty
nice performance benefit in some cases.

joshbetz committed:
API: System Status: Cache theme info

Generating theme info involves reading the theme files from disk and
parsing them.

In testing, caching this data in a transient improves performance by
600ms. Since themes changes relatively infrequently, the cache hit rate
should be quite high.

We're registering the cache clean function in Server.php because
rest_api_init is too late.

joshbetz commented:
If we feel like the additional caching is too risky, I'm happy to remove it. In practice I think only loading the fields that we ask for in `_fields` is going to be a bigger impact. Anecdotally, most requests that the mobile apps make ask specifically for active and inactive plugins.

vedanshujain commented:
hey @joshbetz , general direction of caching expensive results, although unit tests and lining failures need to be addressed.

joshbetz committed:
phpcs

joshbetz committed:
Extract top-level property

Using _fields, we can specify a sub-property like environment.version.
In that case, we need to extract the top-level property so we known
which function to run.

joshbetz committed:
changelog

coreymckrill commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -32,6 +32,30 @@ class WC_REST_System_Status_V2_Controller extends WC_REST_Controller {
+		add_action(
+			'upgrader_process_complete',
+			function( $upgrader, $extra ) {
+				if ( 'theme' !== $extra['type'] ) {
```
Will `$extra['type']` always exist/be set?

coreymckrill replied:
The cached theme data has a couple of props whose values look like they depend on the version of WC (thinking of `has_outdated_templates` and `parent_version_latest`). So it might make sense to clear the cache when WC (and maybe extensions) get updated as well.

coreymckrill commented on plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-system-status-v2-controller.php:
```diff
@@ -1013,93 +1066,107 @@ public function get_dropins_mu_plugins() {
+			);
+
+			$theme_info = array_merge( $active_theme_info, $parent_theme_info );
+			set_transient( 'wc_system_status_theme_info', $theme_info );
```
Some people might modify their theme without triggering an update process. There's probably not a good way to detect that, so instead we could set an expiration here. Maybe a day or a week?

coreymckrill reviewed (commented):
👍 Definitely seeing an improvement when requesting the theme field (~400ms down to ~150ms in my wp-env instance).

Would any of the other fields benefit from being cached?

It seems like it might be a good idea to add some unit tests for potential points of failure here. A couple I'm thinking of at the moment:

* Test that the theme cache actually gets cleared when any of the relevant action hooks gets triggered.
* A nice-to-have would be testing to ensure that `get_item_mappings_per_fields` only returns the requested items.

joshbetz committed:
Clear the theme cache if woocommerce is updated

joshbetz committed:
Expire theme cache in case a theme is updated outside WordPress

If someone updates their theme outside of the WordPress update process
(by manipulating files directly) the cache won't be invalidated. To deal
with that, we're setting the TTL here to 1hr.

joshbetz committed:
Clear the theme cache for any plugin update

The cached theme data has some properties that depend on WooCommerce.
This could also extend to extensions. Since we don't have a reliable way
to know if a given plugin is a WooCommerce extension, any plugin update
should flush the cache.

joshbetz committed:
lint
